text/0000-threadsafe-callback.md

```diff
@@ -0,0 +1,123 @@
+
+impl Drop for ThreadSafeCallbackInner {
+    fn drop(&mut self) {
+        // free resources
+    }
+}
+
+#[derive(Clone)]
+pub struct ThreadSafeCallback(Arc<ThreadSafeCallbackInner>);
+
+impl ThreadSafeCallback {
+    pub fn new<T: Value>(this: Handle<T>, function: Handle<JsFunction>) -> Self;
+    pub fn call<F>(&self, closure: F)
+        where F: FnOnce(&mut TaskContext, Handle<JsValue>, Handle<JsFunction>),
+              F: Send + 'static;
+}
+```
+
+The callback struct is backed by an `std::sync::Arc`. This allows the callback to be clone able and do be send across threads. The arc contains an opaque structure which is used to call the underlying C++ implementation.   
+The C++ implementation is thread safe and handles the asynchronous call to the main thread and the v8 stack setup.   
+Once the last clone of the `ThreadSafeCallback` is out of scope the `Arc` will drop the `ThreadSafeCallbackInner` which in turn will free any resources allocated by the C++ implementation.
+
+*Coming back to the example above:*   
+
+```rust
+let cb = ThreadSafeCallback::new(this, func);
+```
+`ThreadSafeCallback::new` allocates the C++ implementation, which stashes the provided `this`/`function` pair as well as the current `context` and initializes the async handle of `libuv` (`uv_async_init`). This has to occur on the main thread. After that the callback can be cloned and send across threads. 
+```rust
+// call into javascript
+cb.call(move |context, this, func| {
+    // call the js function
+    func.call(context, this, vec![cx.number(i).upcast()]);
+}
+```
+  
+When calling the `call` method, the given closure will be stashed in a thread safe queue and the main thread will be informed via `uv_async_send`. The main thread will call the function registered during the `uv_async_init` call.   
+After setting up the correct v8 scope every closure in the queue will be called with the current context and the stashed `this`/`function` pair.   
+The closure can now perform the Javascript call with the provided `this` and `func` values.
+Once `cb` goes out of scope the C++ implementation closes the async handle, which can only be done in the main thread, and delete itself.
+
+The `Arc` on the Rust side guarantees that the `close` method of the C++ implementation will be only called once and that no further `call`s to the C++ implementation are possible.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This will introduce some unsafe code and the implementation is tied directly to `libuv`
+as it uses the `uv_async_t` API.
+
+# Rationale and alternatives
+[alternatives]: #alternatives
+
+There are no real alternatives, calling into JavaScript from any (Rust) thread is a useful feature for bindings author.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+- Would it make sense to generalize the API e.g. take a `Vec<JsValue>` instead of a `this`/`function` pair?
+- Is `ThreadSafeCallback` a good name for this? Maybe `ValueCapture`?
```

Comment:
I think `ThreadSafe` should be dropped since it is implied by the project. Maybe `vm::Callback` or similar? Could also go with something with `Wake` in the name to patch the `libuv` terminology.

Would `Waker` be too confusing with the new Rust `Waker` API?

Reply:
Agreed on `ThreadSafe` being redundant.

I think this API would get clearer if the naming suggested the _role_ rather than the _mechanism_ of the callback. In particular, what you're doing here is creating a callback to serve as an _event handler_, and then at some point later scheduling (from Rust) an _event_ to invoke the handler (from JS).

So I'd suggest calling this API `neon::event::EventHandler`, and instead of a `call` method it should have a `schedule` method. Then the code example from the RFC would look like this:

```rust
let mut this = cx.this();
let func = cx.argument::<JsFunction>(0)?;
let handler = EventHandler::new(this, func);
thread::spawn(move || {
    for i in 0..100 {
        // do some work...
        thread::sleep(Duration::from_millis(40));
        // schedule a javascript event
        handler.schedule(move |cx, this, func| {
            // call the js function
            func.call(cx, this, vec![cx.number(i).upcast()]);
        })
    }
});
```

Reply:
@dherman Sounds reasonable to me, I'll try to update the RFC

Reply:
@dherman What are your thoughts on this taking a `this` argument? To me, this violates the expectation of typical javascript. I wouldn't expect a callback to have a specified `this` variable unless I specifically bound it.

Is this required to make `() => {}` captures work?

Reply:
It seems strictly more expressive to allow it to specify a `this` argument, since you could always choose to use the global object.

Idiomatically, I think at least some event handlers take custom `this` bindings other than the global object. For example: https://nodejs.org/docs/latest/api/events.html#events_passing_arguments_and_this_to_listeners

If it's really uncommon to want to pass a `this` parameter, we could have two different constructors, where the "normal" API defaults to binding to the global object: `EventHandler::new(func)` and a slightly more verbose one lets you specify the `this`-binding: `EventHandler::bound(func, this)`.

> Is this required to make `() => {}` captures work?

Not as I understand it. If `func` is an arrow function, then calling `func.call(cx, this, ...)` will just disregard the `this` argument.

Reply:
@dherman No, I just joined. Let's coordinate on slack 👍

Reply:
@dherman I've updated the RFC and integrated some of the feedback.